### ISSUE TITLE
"Failed to execute 'define' on 'CustomElementRegistry'" (14.6.0.alpha1)

### ISSUE THREAD

**Issue by fmhdp** [vaadin/flow#10571]
### Description of the bug / feature
Frontend-UI is completely broken
### Versions:
    - Vaadin / Flow version: 14.6.0.alpha1
    - Java version: 11
    - OS version: OSX 10.15.7
    - Browser: Chrome 89.0.4389.114 (current version)

After updating from 14.4.4 to 14.6.0.alpha1 for testing the new packaged themes, our app's frontend is broken. The browser console reports the following error:

`Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry': the name "vaadin-lumo-styles" has already been used with this registry
    at eval (webpack-internal:///../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/version.js:9:16)
    at Module.../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/version.js (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:2833:1)
    at __webpack_require__ (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:20:30)
    at eval (webpack-internal:///../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/color.js:2:69)
    at Module.../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/color.js (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:2773:1)
    at __webpack_require__ (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:20:30)
    at eval (webpack-internal:///../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/mixins/overlay.js:2:67)
    at Module.../node_modules/.pnpm/@vaadin#vaadin-lumo-styles@1.6.1/node_modules/@vaadin/vaadin-lumo-styles/mixins/overlay.js (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:2785:1)
    at __webpack_require__ (http://localhost:8180/VAADIN/build/vaadin-bundle-bfe7ef28df45ff5f8bfb.cache.js:20:30)
    at eval (webpack-internal:///../node_modules/.pnpm/@vaadin#vaadin-overlay@3.5.1/node_modules/@vaadin/vaadin-overlay/theme/lumo/vaadin-overlay-styles.js:2:102)`

This hints at a problem with the "vaadin-lumo-styles" package.
Running "npx find-duplicate-dependencies" gives the following output:
```
This package has the following duplicate dependencies:

@vaadin/vaadin-lumo-styles :
[
  {
    name: '@vaadin/vaadin-lumo-styles',
    version: '1.6.0',
    from: undefined,
    path: 'no-name/@vaadin/vaadin-accordion'
  },
  {
    name: '@vaadin/vaadin-lumo-styles',
    version: undefined,
    from: undefined,
    path: 'no-name/@vaadin/vaadin-core-shrinkwrap'
  }
]
```

Looking into the pnpm-lock.yaml und searching for "vaadin-lumo-styles" reveals 34 results, of which 32 are version 1.6.0 and two are 1.6.1:

```
/@vaadin/vaadin-core-shrinkwrap/14.6.0-alpha1:
    resolution: {integrity: sha512-tq7GH9GgQsGDJ2MrpD5aryy/JdBrMHEbJoSo4Gv+mmweO4VJIffh4QPzLqSaEYeTxaFgq2jxwbGZHpQrBuMzFg==}
    dependencies:
     [...]
      '@vaadin/vaadin-lumo-styles': 1.6.1
     [...]
```
```
  /@vaadin/vaadin-overlay/3.5.1:
    resolution: {integrity: sha512-0g+poK/BXF92L2lSKrHMY5rcKzUxCBZNzP/NDwgi4a86nbjL7CAKKZdno7Yl+j8UsTR76nOEw4fAYTFi86B0qg==}
    dependencies:
      [...]
      '@vaadin/vaadin-lumo-styles': 1.6.1
      [...]
```

I'm anything but knowledgeable around the internal workings and the build process of Vaadin's frontend, but this looks like a dependency problem to me.

I have tried to delete all the config related to the frontend build (package.json, pnpm-lock.yaml, pnpmfile.js, webpack.config.js, webpack.generated.js, node_modules folder) and do a clean-package in maven, to no avail.
Our app has only a single @JsModule annotation and the package loaded this way has a very short list of dependencies (pixi.js, typescript) that are not related to anything Vaadin or Polymer at all.

I say thanks in advance for any help resolving this issue :-)

**Comment by fmhdp** [comment]
Issue #10572 seems to have the same underlying cause, although that one is for an alpha version of Vaadin 20

**Comment by rokkolesa** [comment]
We bumped into the same issue on Vaadin version `14.5.1` using `pnpm`. Which version of `pnpm` were you using? It seems that the `pnpm@6.x` versions have changed the name of hooks file from `pnpmfile.js` to `pnpmfile.cjs`: https://github.com/pnpm/pnpm/releases/tag/v6.0.0

Vaadin always generates `pnpmfile.js` so it doesn't get picked up and no vaadin dependencies get forced versions from `versions.json` resulting in duplicate `vaadin-lumo-styles`.

We resolved the issue by pinning our global `pnpm` version to `5.18.9` (using `npm -g i pnpm@5.18.9`). The app started to work then. We also tried to just copy `pnpmfile.js` to `pnpmfile.cjs` with no success..

Of course if you are using `pnpm < 6.x`, then I suppose it is some other problem.

**Comment by fmhdp** [comment]
That's a good question. I'm relying on whatever Vaadin does and never installed pnpm myself. But I guess you found the root cause. With regular npm, the breakage is gone. Thank you very much for that hint :-)

Apologies in advance if I happened to choose the wrong branch in the flow repository for the source links, but it at least seems to be same code I see In IntelliJ after decompilation.

I tried looking into where Vaadin does invoke pnpm and it seems to be in the FrontendTools: https://github.com/vaadin/flow/blob/6df42e7fe3e9dff63bd299f9f74d0fe2f03f6445/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java

There is even a method called validatePnpmVersion: https://github.com/vaadin/flow/blob/6df42e7fe3e9dff63bd299f9f74d0fe2f03f6445/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java#L625
It only checks for the version being newer than the values defined at the top of the class. And in fact I can see version 6.0.1 appear at this place in the debugger.  
That's expected, though, since getSuitablePnpm first tries plain "pnpm" before trying to use "pnpm@5.15.1", which of course returns the current version.
https://github.com/vaadin/flow/blob/6df42e7fe3e9dff63bd299f9f74d0fe2f03f6445/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java#L607

I guess a short term fix, before supporting pnpm version 6, would be to add a check for whether the version is greater than 5 and forcing version 5 explicity in this case, too.
